File: java-algorithm/src/main/java/com/leetcode/Problem917.java
package com.leetcode;

/**
 * 917. Reverse Only Letters
 * https://leetcode.com/problems/reverse-only-letters/
 */
public class Problem917 {
    public String reverseOnlyLetters(String s) {
        char[] chars = s.toCharArray();
        int i = 0, j = s.length() - 1;
        while (i < j) {
            while (i < j && !Character.isAlphabetic(chars[i])) {
                i++;
            }
            while (i < j && !Character.isAlphabetic(chars[j])) {
                j--;
            }
            if (i < j) {
                char tmp = chars[i];
                chars[i] = chars[j];
                chars[j] = tmp;
                i++;
                j--;
            }
        }
        return String.valueOf(chars);
    }
}

File: python-algo/leetcode/problem_917.py
"""917. Reverse Only Letters
https://leetcode.com/problems/reverse-only-letters/
"""


def reverse_only_letters(s: str) -> str:
    left, right = '', ''
    i, j = 0, len(s) - 1
    while i < j:
        if not s[i].isalpha():
            left += s[i]
            i += 1
            continue
        if not s[j].isalpha():
            right = s[j] + right
            j -= 1
            continue
        left += s[j]
        right = s[i] + right
        i += 1
        j -= 1
    if i == j:
        left += s[i]
    return left + right
